doc/rst/tools/chapel-py/chapel-py.rst

```diff
@@ -0,0 +1,481 @@
+  (``PlusNode(IntNode(1), IntNode(1))``).
+
+The ``chapel.replace`` module provides a driver function ``run``, which, when given
+a source of changes (described below), takes over the execution of the program.
+It registers and parses command line arguments, and handles file modifications.
+Thus, given an AST traversal, you end up with a fully-featured command-line
+script ready to be pointed at files.
+
+.. code-block::
+
+   usage: replace [-h] [--suffix SUFFIX] [--in-place] [filenames ...]
+
+   A tool to search-and-replace Chapel expressions with others
+ 
+   positional arguments:
+     filenames
+ 
+   options:
+     -h, --help       show this help message and exit
+     --suffix SUFFIX
+     --in-place`
+
+
+The 'source of changes' is a Python iterator that should accept two arguments:
+``rc`` (for 'replacement context') and ``root`` (for the AST node at which the
+traversal begins). A very simple (albeit completely pointless and semantically
+incorrect) example is the following script, which replaces all occurrences of
+the number ``42`` with ``meaningOfLife``:
+
+.. code-block:: python
+
+   def replace_all_42(rc, root):
+       for (num, _) in each_matching(root, IntLiteral):
+           if num.text() == '42': yield (num, 'meaningOfLife')
+
+   run(replace_all_42)
+
+A more practical example is renaming methods. Note that this is not
+amenable to naive substitution: we don't want to accidentally rename non-methods
+that happen to have the same name, or even locally declared functions inside
+other functions in a record or class. Consider renaming the ``enterThis``
+method to ``enterContext`` (this is a real change that was performed in
+the Chapel language in 1.32, though it was not done automatically with the
+help of this tool). In the following snippet, only the ``proc`` labeled ``(1)``
+should be renamed. This is because ``enterThis`` is actually a method. On
+the other hand, ``(2)``, even though it's declared within ``R``'s curly braces,
+is actually a locally defined function, and not a method on ``R``.
+``(3)`` is a freestanding function, and thus should also not be renamed.
+
+.. code-block:: chapel
+
+   record R {
+       proc enterThis() {} // (1)
+       proc f() {
+           proc enterThis() {} // (2)
+       }
+   }
+   proc enterThis() {} // (3)
+
+This can be implemented using the following iterator:
+
+.. code-block:: python
+
+   def replace_enter_this(rc, root):
+       for (fn, _) in each_matching(root, Function):
+           if fn.name() == "enterThis" and fn.is_method():
+               yield (fn, "enterContext")
+
+In this case, instead of yielding a pair of the node-to-replace and the
+replacement text, the iterator produces a ``lambda``, which will be fed the
+node's current string representation (``proc enterThis() {...``). Using
+this lambda, we perform a simple substitution, renaming ``enterThis`` to
+``enterContext`` with Python's regular ``replace`` method.
```

Comment:
What happened here? There's no lambda being fed.

Should this code be:

```Python
   def replace_enter_this(rc, root):
       for (fn, _) in each_matching(root, Function):
           if fn.name() == "enterThis" and fn.is_method():
               yield (fn, lambda txt: txt.replace("enterThis", "enterContext", 1))
```

Reply:
Lost in translation from markdown to rst, not sure how that happened...